### PR TITLE
gnrc: add netopt_t to string map

### PR DESCRIPTION
**Pull request by kaspar030**

### CODE REVIEW

**Comment by OlegHahm** [comment]
How about something like this: http://stackoverflow.com/questions/9907160/how-to-convert-enum-names-to-string-in-c

File: sys/net/crosslayer/ng_netapi/ng_netconf_debug.c [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ * @ingroup     net_ng_netapi
+ * @file
+ * @brief       This file contains functionality to map netconf option
+ *              numbers to strings
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ * @}
+ */
+#include "net/ng_netconf.h"
+
```
Comment (authmillenon) [inline sys/net/crosslayer/ng_netapi/ng_netconf_debug.c]: C++ guards are missing, which is very important here due to not using C++ syntax below.

**Comment by kaspar030** [comment]
> How about something like this: http://stackoverflow.com/questions/9907160/how-to-convert-enum-names-to-string-in-c

Hm. I don't care, but I don't want redo this PR. ;)

**Comment by kaspar030** [comment]
- addressed @authmillenon's comment

**Comment by authmillenon** [comment]
I'd give my ACK for this (I would prefer a more scalable solution though, but the solution proposed by @OlegHahm is not it, IMHO)

**Comment by kaspar030** [comment]
- squashed

**Comment by authmillenon** [comment]
Causes (valid) build errors.

**Comment by kaspar030** [comment]
Ah, it's missing ```NETCONF_OPT_DEVICE_TYPE```.

@authmillenon, gnrc uses ```NETCONF_OPT_PROTO``` to distinguish device types, with protocols being sixlowpan, ipv6, icmpv6. Wouldn't it be more to the point if either a device returns protocols like ethernet, 802.15.4, cc110x, or we define ```NETCONF_OPT_DEVICE_TYPE``` returning those?

The rationale is that e.g., an ethernet device can support multiple protocols (ipv4+ipv6), but when using ```NETCONF_OPT_PROTO``` as is, it can only reply with one. Also, IMHO link-layer header should have the right ```PROTO``` set.

**Comment by authmillenon** [comment]
IMHO this belongs into a seperate issue or the devel mailing list for discussion, but I'll answer it anyways: `NETCONF_OPT_PROTO` is supposed to be fallback solution if and only if the link-layer protocol does not support multiplexing. Link-layers like e. g. the ethertype in ethernet usually have a field that identifies the network layer. (Otherwise how do you tell which protocol to send to?) Maybe we should update the doc on that. 

As for `NETCONF_OPT_DEVICE_TYPE`: I really don't see a use-case for that, but if you do, I would prefer to use this new option type instead of changing semantics of existing options around to get at least some stability in the API. 

Funny enough, the original purpose for `NETCONF_OPT_PROTO` was what you proposed `NETCONF_OPT_DEVICE_TYPE` to be for, until the link-layer did not get a nettype anymore causing @haukepetersen to repurpise the option as described above.

**Comment by kaspar030** [comment]
> As for `NETCONF_OPT_DEVICE_TYPE`: I really don't see a use-case for that

Well, e.g. IPv6 or sixlowpan could enumerate interfaces, and if it has a
known device type, use it accordingly. Right know we're asking for
"PROTO", and if it is "SIXLOWPAN", sixlowpan will use it. Devices
shouldn't know about upper layers.

I would also go for "_TYPE" instead of reusing "_PROTO", as "_PROTO" is
used to describe the format of data (link_layer header type), and not
the semantics of a device. (PROTO makes more sense for upper layers,
anyway).

> Link-layers like e. g. the ethertype in ethernet usually have a field
that identifies the network layer. (Otherwise how do you tell which
protocol to send to?)

Once a packet arrives at the link layer from an upper layer, the link
layer should set it's "etherype" depending in the pktsnip->type field.
But in order to send packets to this link layer in the first place, the
upper layer needs to know that it can do so. IMHO,
"NETCONF_OPT_DEVICE_TYPE" is the way to go here.

If you agree, I'd PR the NETCONF_OPT_DEVICE_TYPE addition.

**Comment by authmillenon** [comment]
> Well, e.g. IPv6 or sixlowpan could enumerate interfaces, and if it has a
known device type, use it accordingly. Right know we're asking for
"PROTO", and if it is "SIXLOWPAN", sixlowpan will use it. Devices
shouldn't know about upper layers.

The situation we face (even now already) that there are vastly more (especially proprietary) link-layers than network layers (where it makes sense to write a `gnrc` module that is) and especially adaption-layers (like 6LoWPAN). So it seems far simpler (or code size saving for that matter) to me that instead of checking a long list of possible `enum` values the device to just say "Japp, I'm a 6LoWPAN device, please give me packets of that type".

**Comment by authmillenon** [comment]
But if you think this way around it would be cleaner, please open a PR and let's discuss it there.

File: sys/include/net/ng_netconf.h [hunk 2]
```diff
@@ -177,6 +177,20 @@ typedef enum {
     /* add other states if needed */
 } ng_netconf_state_t;
 
+/**
+ * @brief   Map of netconf options to strings, for debugging
+ */
+extern const char * ng_netconf_strmap[];
```
Comment (haukepetersen) [inline sys/include/net/ng_netconf.h]: why would you make this public?

**Comment by haukepetersen** [comment]
Regarding the `NETCONF_OPT_PROTO`: as @authmillenon stated, this option was introduced for link-layers, that don't support multiplexing of protocols. The main use case here is actually when receiving data: say you have a radio that receives a packet, how does it know where who to hand over this packet? So do you hand it to 6LoWPAN, or IPv4 or IPv6? By setting the PROTO option, you can basically configure, what kind of data you send on that link...

**Comment by kaspar030** [comment]
So in this case we should change proto into a bitfield. When any layer get's a 'proto' get call, it replies with all the protocols it supports (e.g., can handle). Additionally we add the raw stuff (ethernet, 802.15.4) to valid values, as link-layers might just support that.

E.g., sixlowpan can then query a device, and if it sees PROTO_802.15.4 in it's protocol list, use it.

When receiving data, the link layer uses the same "PROTO" values to mark payload.

I would still add the "TYPE" option, so upper layers can query the device type and handle it if necessary.

**Comment by kaspar030** [comment]
- rebased
- not exporting _ng_netconf_strmap anymore
- handle case if an option is not part of the array

**Comment by authmillenon** [comment]
Needs to be adapted to changes in #3457

**Comment by authmillenon** [comment]
Ping

**Comment by authmillenon** [comment]
Please squash. ACK, when travis is happy.

File: sys/net/crosslayer/ng_netapi/netconf_debug.c [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+ * @}
+ */
+
+#include "net/netopt.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+static const char * _netopt_strmap[] = {
```
Comment (authmillenon) [inline sys/net/crosslayer/ng_netapi/netconf_debug.c]: cosmetics: `char *_netopt_strmap[]`

File: sys/net/crosslayer/ng_netapi/netconf_debug.c [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    [NETOPT_RAWMODE] = "NETOPT_RAWMODE",
+    [NETOPT_RX_START_IRQ] = "NETOPT_RX_START_IRQ",
+    [NETOPT_RX_END_IRQ] = "NETOPT_RX_END_IRQ",
+    [NETOPT_TX_START_IRQ] = "NETOPT_TX_START_IRQ",
+    [NETOPT_TX_END_IRQ] = "NETOPT_TX_END_IRQ",
+    [NETOPT_AUTOCCA] = "NETOPT_AUTOCCA",
+    [NETOPT_NUMOF] = "NETOPT_NUMOF",
+};
+
+const char * netopt2str(netopt_t opt)
```
Comment (authmillenon) [inline sys/net/crosslayer/ng_netapi/netconf_debug.c]: here too

File: sys/net/crosslayer/ng_netapi/netconf_debug.c [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+{
+    if (opt >= NETOPT_NUMOF) {
+        return "unknown";
+    }
+    else {
+        const char * tmp = _netopt_strmap[opt];
```
Comment (authmillenon) [inline sys/net/crosslayer/ng_netapi/netconf_debug.c]: here too

File: sys/include/net/netopt.h [hunk 6]
```diff
@@ -177,6 +177,15 @@ typedef enum {
...
 } netopt_state_t;
 
+/**
+ * @brief   Get a string ptr corresponding to opt, for debugging
+ *
+ * @param[in] opt   The option to get a string representation for
+ *
+ * @return          ptr to string representation for given option or "unknown"
+ */
+const char * netopt2str(netopt_t opt);
```
Comment (authmillenon) [inline sys/include/net/netopt.h]: here too

**Comment by authmillenon** [comment]
You can leave the cosmetics as is, we need to clean up the code anyways :-)

**Comment by kaspar030** [comment]
- rebased
- moved to gnrc directory
- fixed whitespaces

@authmillenon do you agree with the new location?

**Comment by authmillenon** [comment]
> @authmillenon do you agree with the new location?

So your stack should not use it, neither should `netdev2`? I would prefer creating a new `gnrc`-indepent module for that at `sys/net/crosslayer/netopt/`.

File: sys/net/crosslayer/ng_netapi/netconf_debug.c [hunk 7]
```diff
@@ -0,0 +1,65 @@
+/*
```
Comment (authmillenon) [inline sys/net/crosslayer/ng_netapi/netconf_debug.c]: Shouldn't this file not exist anymore?

Comment (kaspar030) [inline sys/net/crosslayer/ng_netapi/netconf_debug.c]: correct, my bad.

**Comment by kaspar030** [comment]
> So your stack should not use it, neither should netdev2?

True, netdev2 needs this. Will find a better place.

**Comment by authmillenon** [comment]
> True, netdev2 needs this. Will find a better place.

Just make a `netopt` module. It will only be added to the binary if it's objects are used anyway.

**Comment by kaspar030** [comment]
- rebased
- addressed @authmillenon's comment
- added unittest

**Comment by kaspar030** [comment]
- proactively squashed

**Comment by authmillenon** [comment]
ACK and go.

**Pull request merged**